daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala

```diff
@@ -34,42 +36,38 @@ final class TransactionBuilder(pkgTxVersion: Ref.PackageId => TransactionVersion
 
   def add(node: Node): NodeId = ids.synchronized {
     val nodeId = newNode(node)
-    roots += nodeId
+    roots = roots :+ nodeId
     nodeId
   }
 
-  def add(node: Node, parent: NodeId): NodeId = ids.synchronized {
+  def add(node: Node, parentId: NodeId): NodeId = ids.synchronized {
     lazy val nodeId = newNode(node) // lazy to avoid getting the next id if the method later throws
-    nodes = nodes.mapResult { ns =>
-      val VersionedNode(version, node) = ns(parent)
-      val exercise = node match {
-        case exe: TxExercise => exe
-        case _ =>
-          throw new IllegalArgumentException(
-            s"Node ${parent.index} either does not exist or is not an exercise")
-      }
-      ns.updated(
-        parent,
-        VersionedNode(
-          version,
-          exercise.copy(children = exercise.children.slowSnoc(nodeId))
-        ))
+    nodes(parentId) match {
+      case VersionedNode(_, _: TxExercise) =>
+        children += parentId -> (children(parentId) :+ nodeId)
+      case _ =>
+        throw new IllegalArgumentException(
+          s"Node ${parentId.index} either does not exist or is not an exercise")
     }
     nodeId
   }
 
   def build(): Tx.Transaction = ids.synchronized {
     import VersionTimeline.Implicits._
-    val builtNodes = nodes.result()
-    val builtRoots = roots.result()
+    val builtNodes = nodes.transform {
+      case (nid, VersionedNode(version, exe: TxExercise)) =>
+        VersionedNode[NodeId, ContractId](version, exe.copy(children = children(nid).toImmArray))
+      case (_, node) =>
+        node
+    }
     val nodesVersions =
-      builtRoots.reverseIterator
+      roots.reverseIterator
```

Comment:
I imagine iterating over a `BackStack` from either side to be more expensive than iterating over an `ImmArray`. If that's the case, we should convert to an `ImmArray`, which we are doing anyway, first and then iterate over that.

Reply:
`Backstack` has only `reverseIterator`. I think this is pretty efficient. 
Anyway, as we convert roots to an `ImmArray`, just below I applied your suggestion.